consensus/src/dag/payload/payload_fetcher.rs

```diff
@@ -0,0 +1,253 @@
+            request: PayloadRequest::from(metadata),
+            res_tx,
+            responders,
+        };
+        self.command_tx
+            .try_send(request)
+            .map_err(|e| anyhow::anyhow!("unable to send request {}", e))?;
+        Ok(res_rx)
+    }
+
+    pub fn cancel(&self, metadata: PayloadInfo) -> anyhow::Result<()> {
+        self.command_tx
+            .try_send(FetchServiceCommand::CancelFetch(metadata))
+            .map_err(|_| anyhow::anyhow!("unable to send cancel"))?;
+        Ok(())
+    }
+}
+
+pub struct PayloadFetcherService {
+    inner: Arc<PayloadFetcher>,
+    payload_store: Arc<DagPayloadStore>,
+    command_rx: Receiver<FetchServiceCommand>,
+    futures: FuturesUnordered<Pin<Box<dyn Future<Output = Result<PayloadDigest, Aborted>> + Send>>>,
+    inprogress_reqs: HashMap<PayloadDigest, AbortHandle>,
+}
+
+impl PayloadFetcherService {
+    pub fn new(
+        epoch_state: Arc<EpochState>,
+        network: Arc<dyn TDAGNetworkSender>,
+        payload_store: Arc<DagPayloadStore>,
+        time_service: TimeService,
+        config: DagFetcherConfig,
+    ) -> (Self, PayloadRequester) {
+        let (command_tx, command_rx) = tokio::sync::mpsc::channel(100);
+        (
+            Self {
+                inner: Arc::new(PayloadFetcher {
+                    epoch_state,
+                    network,
+                    time_service,
+                    config,
+                }),
+                payload_store,
+                command_rx,
+                futures: FuturesUnordered::new(),
+                inprogress_reqs: HashMap::new(),
+            },
+            PayloadRequester { command_tx },
+        )
+    }
+
+    pub async fn start(self) {
+        let Self {
+            inner,
+            payload_store,
+            mut command_rx,
+            mut futures,
+            mut inprogress_reqs,
+        } = self;
+        loop {
+            select! {
+                Some(result) = futures.next() => {
+                    if let Ok(digest) = result {
+                        inprogress_reqs.remove(&digest);
+                    }
+                },
+                // TODO: configure limit concurrent futures
+                Some(command) = command_rx.recv(), if futures.len() < 50 => {
+                    match command {
+                        FetchServiceCommand::Request{request, responders, res_tx} => {
+                            let id = request.id();
+                            let digest = *request.payload_digest();
+
+                            if !payload_store.is_missing(id, &digest) {
+                                debug!("payload already exists: {:?}", request);
+                                continue;
+                            }
+
+                            if inprogress_reqs.contains_key(&digest) {
+                                debug!("payload already requested {:?}", request);
+                                continue;
```

Comment:
You are droping the tx and the caller will never get the payload.
You probably should register all duplicated call and send back payload to all once received

Reply:
+1 to the dropped tx here